server/api/solar_lottery.go

```diff
@@ -0,0 +1,392 @@
+			}
+		}
+	}
+
+	// Get a snapshot of the initial probability distribution
+	var cdf *cdf
+	var unmet []store.Need
+	var pickedUser, chosenUser *User
+	var qualifiedUsers UserMap
+	cycle := func() {
+		if pickedUser != nil {
+			delete(pool, pickedUser.MattermostUserID)
+			delete(qualifiedUsers, pickedUser.MattermostUserID)
+		}
+		if chosenUser != nil {
+			chosen[chosenUser.MattermostUserID] = chosenUser
+			chosenUser = nil
+		}
+
+		cdf = api.calculateCDF(rotation, pool, shiftNumber)
+		rotation.sortNeedsByHeat(cdf)
+		unmet = unmetNeeds(rotation.Needs, chosen)
+	}
+
+	// Set and snapshot the initial state for logging later
+	cycle()
+
+	out := fmt.Sprintf("filling %s, choosing from:\n", MarkdownShift(rotation, shiftNumber))
+	sort.Sort(cdf)
+	for i, id := range cdf.ids {
+		out += fmt.Sprintf("- **%.3f**: %s\n", cdf.weights[i]/cdf.total, api.MarkdownUserWithSkills(pool[id]))
+	}
+	api.Logger.Debugf("%s", out)
+
+CYCLE:
+	for ; len(unmet) > 0; cycle() {
+		need := unmet[0]
+		autoerr := func(err error) autofillError {
+			return autofillError{
+				causeUnmetNeed:  &need,
+				causeUnmetNeeds: unmet,
+				causeCapacity:   rotation.Size - len(chosen),
+				orig:            err,
+				shiftNumber:     shiftNumber,
+			}
+		}
+
+		if need.Max == 0 {
+			// nothing else to do, not a realistic possibility unless misconfigured
+			continue
+		}
+
+		if len(chosen) >= rotation.Size {
+			return autoerr(ErrFailedSizeExceeded)
+		}
+
+		qualifiedUsers = usersQualifiedForNeed(pool, need)
+		if len(qualifiedUsers) == 0 {
+			return autoerr(ErrFailedInsufficientForNeeds)
+		}
+
+		for ; len(qualifiedUsers) > 0; cycle() {
+			pickedUser = api.pickUser(rotation, qualifiedUsers, shiftNumber)
+			if pickedUser == nil {
+				// not really a possibility
+				return autoerr(errors.Errorf("failed to pick a user out of %v qualified", len(qualifiedUsers)))
+			}
+
+			var maxedNeed *store.Need
+			for _, rotationNeed := range rotation.Needs {
+				if rotationNeed.Max < 0 {
+					continue
+				}
+				if qualifiedForNeed(pickedUser, rotationNeed) &&
+					rotationNeed.Max-len(usersQualifiedForNeed(chosen, rotationNeed))-1 < 0 {
+					maxedNeed = &rotationNeed
+					break
+				}
+			}
+
+			if maxedNeed != nil {
+				api.Logger.Debugf("Disqualified user %s from %s: would hit max on %s.",
+					api.MarkdownUser(pickedUser), MarkdownNeed(need), MarkdownNeed(*maxedNeed))
+				continue
+			}
+
+			chosenUser = pickedUser
+			api.Logger.Debugf("Accepted user %s for %s", api.MarkdownUser(chosenUser), MarkdownNeed(need))
+			continue CYCLE
+		}
+	}
+
+	// All needs are met
+
+	autoerr := func(err error) autofillError {
+		return autofillError{
+			causeCapacity: rotation.Size - len(chosen),
+			orig:          err,
+			shiftNumber:   shiftNumber,
+		}
+	}
+
+	for ; len(chosen) < rotation.Size && len(pool) > 0; cycle() {
+		// Choose next user from the pool, weighted-randomly.
+		pickedUser = api.pickUser(rotation, pool, shiftNumber)
```

Comment:
Possibly a helper function between here and 139-167. Pick users from pool? Might be worth it for testing. 0/5

Reply:
y, I lazily/purposely left them separate while debugging the initial implementation, but yes, will consolidate.